rclcpp/include/rclcpp/detail/qos_parameters.hpp

```diff
@@ -0,0 +1,303 @@
+
+/// \internal Modify the given `policy` in `qos` to be `value`.
+inline
+void
+apply_qos_override(
+  rclcpp::QosPolicyKind policy, rclcpp::ParameterValue value, rclcpp::QoS & qos);
+
+/// \internal Declare qos parameters for the given entity.
+/**
+ * \tparam NodeT Node pointer or reference type.
+ * \tparam EntityQosParametersTraits A class with two static methods: `entity_type()` and
+ *  `allowed_policies()`. See `PublisherQosParametersTraits` and `SubscriptionQosParametersTraits`.
+ * \param options User provided options that indicate if qos parameter overrides should be
+ *  declared or not, which policy can have overrides, and optionally a callback to validate the profile.
+ * \param node Parameters will be declared using this node.
+ * \param topic_name Name of the topic of the entity.
+ * \param default_qos User provided qos. It will be used as a default for the parameters declared.
+ * \return qos profile based on the user provided parameter overrides.
+ */
+template<typename NodeT, typename EntityQosParametersTraits>
+std::enable_if_t<
+  rclcpp::node_interfaces::has_node_parameters_interface<
+    decltype(std::declval<typename rcpputils::remove_pointer<NodeT>::type>())>::value ||
+  std::is_same<typename std::decay_t<NodeT>,
+  rclcpp::node_interfaces::NodeParametersInterface::SharedPtr>::value,
+  rclcpp::QoS>
+declare_qos_parameters(
+  const ::rclcpp::QosOverridingOptions & options,
+  NodeT & node,
+  const std::string & topic_name,
+  const ::rclcpp::QoS & default_qos,
+  EntityQosParametersTraits)
+{
+  auto & parameters_interface = *rclcpp::node_interfaces::get_node_parameters_interface(node);
+  std::string param_prefix;
+  {
+    std::ostringstream oss{"qos_overrides.", std::ios::ate};
+    oss << topic_name << "." << EntityQosParametersTraits::entity_type();
+    if (!options.id.empty()) {
+      oss << "_" << options.id;
+    }
+    oss << ".";
+    param_prefix = oss.str();
+  }
+  std::string param_description_suffix;
+  {
+    std::ostringstream oss{"} for ", std::ios::ate};
+    oss << EntityQosParametersTraits::entity_type() << " {" << topic_name << "}";
+    if (!options.id.empty()) {
+      oss << " with id {" << options.id << "}";
+    }
+    param_description_suffix = oss.str();
+  }
+  rclcpp::QoS qos = default_qos;
+  for (auto policy : EntityQosParametersTraits::allowed_policies()) {
+    if (
+      std::count(options.policy_kinds.begin(), options.policy_kinds.end(), policy))
+    {
+      std::ostringstream param_name{param_prefix, std::ios::ate};
+      param_name << qos_policy_kind_to_cstr(policy);
+      std::ostringstream param_desciption{"qos policy {", std::ios::ate};
+      param_desciption << qos_policy_kind_to_cstr(policy) << param_description_suffix;
+      rcl_interfaces::msg::ParameterDescriptor descriptor{};
+      descriptor.description = param_desciption.str();
+      descriptor.read_only = true;
+      auto value = parameters_interface.declare_parameter(
```

Comment:
I noticed that this can cause a [parameter already declare](https://github.com/ros2/rclcpp/blob/371074523adfeeb69479266a044a7a8a34712efa/rclcpp/include/rclcpp/exceptions/exceptions.hpp#L258) error.

When creating two subscriptions in the same topic, the same qos might be desired for both (i.e. not passing an id to disambiguate) , and in that case the second creation will fail.
Maybe the `QosOverridingOptions` should have a "parameters already declared" option, that uses `get_parameter` instead of `declare_parameter`.

"/clock" is an special case, because we can create many time sources with different subscriptions, and we don't know what's the first one (I think that never happens in a real use case, but it's happening in tests).
[Here](https://github.com/ros2/rclcpp/blob/371074523adfeeb69479266a044a7a8a34712efa/rclcpp/src/rclcpp/time_source.cpp#L85-L92) is how we handle that for `use_sim_time` (by the way, there's a TOCTTOU race condition there).
I'm not sure what's the right way to handle this last case, @jacobperron @wjwwood any ideas?

Reply:
What about wrapping this in a try-catch and if we catch a `ParameterAlreadyDeclaredException` then use `get_parameter`?

Reply:
Perhaps the race you found can be fixed in a similar way.

Reply:
> What about wrapping this in a try-catch and if we catch a ParameterAlreadyDeclaredException then use get_parameter?

That sounds ok to me, but maybe in the general case it should be an error (the `/clock` case sounds pretty exceptional).

> Perhaps the race you found can be fixed in a similar way.

:+1:

Reply:
I guess we still need to resolve this. In the general case, how should users workaround this issue?

IIUC, the situation arise if I do something like this:

```c
...
 rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};
 auto sub_options = rclcpp::SubscriptionOptions();
 sub_options.qos_overriding_options = qos_overrides;
 auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", foo_qos, foo_callback, sub_options);
 auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", bar_qos, bar_callback, sub_options);
...
```
At first glance, I would expect things to work without issue; the user shouldn't have to worry about parameters already being declared since it's an implementation detail. But, then it looks a little confusing since `sub1` and `sub2` have different "default" QoS settings, but the same override abilities. And configuring one QoS policy should modify both. It seems a bit strange, and is probably a exceptional use-case, but seems like something we could easily support.

Correct me if I'm wrong or missing something.

Reply:
mmm, the default qos of the second case will be ignored, so I don't think allowing that is a good idea.
We definitely need to support this somehow.

Maybe we should provide user facing `declare_subscription_qos_parameter_overrides` and `get_subscription_qos_parameter_overrides` functions:

```cpp
rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};
rclcpp::QoS qos{KeepLast(10)};
declare_subscription_qos_parameter_overrides(
  qos_overrides, node_parameters, node_topics->resolve_topic_name(topic), qos);  # qos is inout here
auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", qos, foo_callback);
```
 ... in another part of the code later ...
```cpp
rclcpp::QoS qos = get_subscription_qos_from_overrides(
  node_parameters, node_topics->resolve_topic_name(topic));  // qos overriding options aren't needed here
auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", qos, bar_callback, sub_options);
```

The function for declaring the parameters is not strictly necessary, as it can be done in a `create_subscription` call passing qos overriding options. But in some cases, like `/clock`, I think it will be useful.

Reply:
I don't think the user should have to worry about declaring QoS parameters in this context. As Jacob mentioned, this should be a implementation concern. In Navigation 2 we used internally a declare_parameter_if_not_declared function, as this came up in the context of plugins declaring the same parameter: https://github.com/ros-planning/navigation2/blob/main/nav2_util/include/nav2_util/node_utils.hpp#L83-L94.

I would also strive to keep the semantics very clear. In the case of the following example,

```
 rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};

 auto sub_options = rclcpp::SubscriptionOptions();
 sub_options.qos_overriding_options = qos_overrides;

 auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", foo_qos, foo_callback, sub_options);

 auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", bar_qos, bar_callback, sub_options);
```

I would interpret this to mean that the first subscription is using any QoS settings from foo_qos and then overriding those with any settings in sub_options.qos_overriding_options. Similarly, for the second subscription, it will use bar_qos and then override any QoS settings from sub_options.qos_overriding_options. Assuming this is the intent, it is clear what to expect.

In contrast, the second example, which has this:

```
rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};
rclcpp::QoS qos{KeepLast(10)};
declare_subscription_qos_parameter_overrides(
  qos_overrides, node_parameters, node_topics->resolve_topic_name(topic), qos);  # qos is inout here
auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", qos, foo_callback);
```
Then this:

```
rclcpp::QoS qos = get_subscription_qos_from_overrides(
  node_parameters, node_topics->resolve_topic_name(topic));  // qos overriding options aren't needed here
auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", qos, bar_callback, sub_options);
```
(Assuming I've understood correctly) requires the user to declare the QoS parameters, uses separate functions depending on whether it is the first call to get overrides or not, and also seems to have the side effect of setting the QoS state so that get_subscription_qos_from_overrides can apply it. IMO, all of this is more complicated than the first version.

Reply:
Regarding the approach in the current code..

```
 rclcpp::QoS qos{rclcpp::KeepLast(10)};
  qos = rclcpp::detail::declare_qos_parameters(
    rclcpp::QosOverridingOptions::kDefaultPolicies,
    node,
    "/my/fully/qualified/topic_name",
    qos,
    rclcpp::detail::PublisherQosParametersTraits{});

  EXPECT_EQ(
    node->get_parameter(
      "qos_overrides./my/fully/qualified/topic_name.publisher.history").get_value<std::string>(),
    "keep_last");

```
Do we want to expose that this QoS override functionality is implemented using parameters? You could abstract this and expose only the ability to set (and get) QoS override values for a topic name and the ability to apply these overrides to a QoS policy.

Reply:
> I would interpret this to mean that the first subscription is using any QoS settings from foo_qos and then overriding those with any settings in sub_options.qos_overriding_options. Similarly, for the second subscription, it will use bar_qos and then override any QoS settings from sub_options.qos_overriding_options. Assuming this is the intent, it is clear what to expect.

The issue is the following:

```cpp
 auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", foo_qos, foo_callback, sub_options);
```

supposing that `reliability` is configurable, this will declare a parameter named `qos_overrides./same_topic.subscription.reliability` with a default value of `foo_qos.reliability`.

```cpp
 auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", bar_qos, bar_callback, sub_options);
```

This second call will not declare a parameter (as it was already declared before), and `bar_qos.reliability` will be completely ignored and overridden with the value got from the parameter.

I prefer asymmetries being explicit and not implicit, and the API proposed above makes the asymmetry explicit.

We could also have some extra sugar in the `QosOverridingOptions` struct, like an option to indicate if declare parameters or to get the previously declared ones:

```cpp
 rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};

 auto sub_options = rclcpp::SubscriptionOptions();
 sub_options.qos_overriding_options = qos_overrides;

 auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", foo_qos, foo_callback, sub_options);

sub_options.qos_overriding_options.declare_parameters = false;  // the second construction will get the already declared parameters instead of declaring again.

 auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
    node, "same_topic", bar_qos, bar_callback, sub_options);
```

Reply:
> requires the user to declare the QoS parameters, uses separate functions depending on whether it is the first call to get overrides or not, and also seems to have the side effect of setting the QoS state so that get_subscription_qos_from_overrides can apply it

Ah, that's a good point.
An input qos parameters profile would still be needed in `get_subscription_qos_from_overrides` (as having an extra qos state doesn't seem like a good idea).

```cpp
rclcpp::QoS qos{KeepLast{10}};
get_subscription_qos_from_overrides(
  node_parameters, node_topics->resolve_topic_name(topic), qos);
```

Reply:
> Do we want to expose that this QoS override functionality is implemented using parameters? I understood that it could be implemented using parameters, but didn't realize this would be part of the interface.

If we're using parameters, the user will always be able to use the "normal" parameters API to get them.
In this case they cannot set the parameter, as we're using the `read-only` option.

> You could abstract this and expose only the ability to set (and get) QoS override values for a topic name and the ability to apply these overrides to a QoS policy.

IIUC, that's the idea of the previously proposed `get_subscription_qos_from_overrides`, `declare_subscription_qos_parameter_overrides`.

Reply:
IIUC, using the proposed `declare_subscription_qos_parameter_overrides` and `get_subscription_qos_parameter_overrides functions` functions is essentially the same as the user setting the same default QoS for both subscriptions. IMO we should encourage people to set the same defaults for sub/pubs with the same topic name, rather than adding additional functions for them to use. I think this is simpler and easy to document. If possible, it would be good to log a warning if we detect a different default QoS setting in the second subscription being created.

Reply:
As I understand, even if we introduce the proposed functions we still rely on the user knowing about this pitfall. We should probably document this pitfall in the design doc and somewhere in the code API.

Reply:
Ok, I understand that using (read-only) parameters is going to be part of the interface to the user and not a hidden implementation detail. Given that, could the interface be simplified a bit in its operations? The user could perhaps:

- Explicitly declare (once) each parameter that can be overridden using `declare_subscription_qos_parameter_overrides`
- Update an existing QoS using `get_subscription_qos_from_overrides`
- Create the subscription using the update QoS

For example, the previous example:
```
rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};
rclcpp::QoS qos{KeepLast(10)};
declare_subscription_qos_parameter_overrides(
  qos_overrides, node_parameters, node_topics->resolve_topic_name(topic), qos);  # qos is inout here
auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", qos, foo_callback);

rclcpp::QoS qos = get_subscription_qos_from_overrides(
  node_parameters, node_topics->resolve_topic_name(topic));  // qos overriding options aren't needed here
auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", qos, bar_callback, sub_options);
```
Would become:

```
// Declare each override once, which sets the value in the node parameters (without modifying a particular qos)
rclcpp::QosOverridingOptions qos_overrides{rclcpp::QosOverridingOptions::kDefaultPolicies};
declare_subscription_qos_parameter_overrides(
    qos_overrides, node_parameters, node_topics->resolve_topic_name(topic));

// Now, you can override a particular QoS setting, updating any qos fields from the parameters
rclcpp::QoS qos{KeepLast(10)};
rclcpp::QoS updated_qos = get_subscription_qos_from_overrides(
    qos, node_parameters, node_topics->resolve_topic_name(topic));

// Then, create a subscription
auto sub1 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", updated_qos, foo_callback);

// You can override another
rclcpp::QoS qos2{KeepLast(5)};
rclcpp::QoS updated_qos2 = get_subscription_qos_from_overrides(
    qos2, node_parameters, node_topics->resolve_topic_name(topic));

// And create another subscription
auto sub2 = rclcpp::create_subscription<test_msgs::msg::Empty>(
  node, "same_topic", updated_qos2, bar_callback, sub_options);
```

There are basically two operations needed: create a parameter for a QoS override and apply QoS overrides to a QoS structure.

Reply:
> As I understand, even if we introduce the proposed functions we still rely on the user knowing about this pitfall

It doesn't sound like a pitfall in `get_subscription_qos_from_overrides` (or with  `qos_overriding_options.declare_parameters = false`), as in that case the user would be explicitly requesting: "take this qos profile and always replace the policies that have a declared parameter with the parameter value".

> Ok, I understand that using (read-only) parameters is going to be part of the interface to the user and not a hidden implementation detail

Yeah, that was the result of some discussions in https://github.com/ros2/design/pull/296.

> Given that, could the interface be simplified a bit in its operations? The user could perhaps:

That sounds ok to me.
It's a bit too verbose in the case you want to only create one subscription/publisher in that topic, so I wanted to provide extra sugar on top of that to avoid the verbosity.

Reply:
True, it may be more lines of code, but I think the programming model of: Declare, Override, Create, Override, Create is simpler and easier to understand than DeclareAndOverride, Create, OverrideWithSavedQoS, Create

I'm assuming that this operation is returning a saved QoS from the initial DeclareAndOverride (could be wrong about that)

```
rclcpp::QoS qos = get_subscription_qos_from_overrides(
  node_parameters, node_topics->resolve_topic_name(topic));  // qos overriding options aren't needed here
```

Reply:
Here's a straw-man example of what I think we should do here:

```c++
if (!parameters_interface.has_parameter()) {
  auto value = parameters_interface.declare_parameter(
    param_name.str(), get_default_qos_param_value(policy, qos), descriptor);
  ::rclcpp::detail::apply_qos_override(policy, value, qos);
} else {
  auto value = parameters_interface.get_parameter(param_name.str());
  // TODO: We would need to add the "is_default()" API, but it should be easy
  if (value.is_default() && value != get_default_qos_param_value(policy, qos)) {
    // Log or output an error here (optionally, we could throw an exception instead)
    std::cerr << "Default setting for QoS policy " << policy << " is being ignored. Previously overridden by another subscription." << std::endl; 
  }
  ::rclcpp::detail::apply_qos_override(policy, value, qos);
}
```

Reply:
To clarify, I think it is a pitfall in the sense that the user must be aware that they can't write code like my example (https://github.com/ros2/rclcpp/pull/1408#discussion_r514514169), furthermore, they must also be aware of these additional functions to workaround the issue (which are essentially setting the default QoS profile to the same value as a previously declared sub/pub).

Reply:
https://github.com/ros2/rclcpp/pull/1408/commits/cab74d8d2de3238d87f9f0853004ae08cb1cee9f takes the approach of https://github.com/ros2/rclcpp/pull/1408#discussion_r517669584.